lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -220,112 +229,177 @@ define((require, exports, module) ->
-        # ending sequence, if it is inside one
-        # The loop isn't executed, if current token is null, because:
-        #   a) we don't need to -- if `tokenIterator` is initially on
-        #      the empty line, then it is guaranteed not to be inside start sequence
-        #   b) it can cause bugs -- if `tokenIterator` is initially after the end of a document,
-        #      current token is null, and after stepping backward `tokenIterator` is on the
-        #      end of a document. If end sequence happens to be there, then equation end is
-        #      then found without any problem, whereas in this case null should be returned
-        for token in EquationRangeHandler.END_EQUATION_TOKEN_SEQUENCE.slice(0).reverse()
-          if EquationRangeHandler.equalTokens(token, tokenIterator.getCurrentToken())
-            tokenIterator.stepBackward()
-
-      curSequenceIndex = 0
-      curEquationStart = null
-      while curSequenceIndex < EquationRangeHandler.END_EQUATION_TOKEN_SEQUENCE.length
-        curToken = tokenIterator.stepForward()
-        if not curToken
+        for boundarySequence in boundarySequences
+          for token in boundarySequence.slice(0).reverse()
+            curToken = tokenIterator.getCurrentToken()
+            if curToken? and EquationRangeHandler.equalTokens(token, curToken)
+              moveFromBoundary()
+
+      allowedSequences = Array(boundarySequences.length)
+      allowedSequences.fill(true)
+      curSequence = null
+      curSequenceId = null
+      curSequenceIndex = null
+      curEquationOuterBoundary = null
+      curEquationInnerBoundary = null
+      while true
+        moveToBoundary()
+        curToken = tokenIterator.getCurrentToken()
+        if not curToken?
           return null
-        if EquationRangeHandler.equalTokens(
-            EquationRangeHandler.END_EQUATION_TOKEN_SEQUENCE[curSequenceIndex],
-            curToken)
-          if curSequenceIndex == 0
-            curEquationStart = tokenIterator.getCurrentTokenPosition()
-          curSequenceIndex += 1
-        else
-          curSequenceIndex = 0
-          curEquationStart = null
-      return curEquationStart
+
+        if curSequence != null
+          if EquationRangeHandler.equalTokens(
+              curSequence[curSequenceIndex],
+              curToken)
+            curSequenceIndex += 1
+          else
+            allowedSequences[curSequenceId] = false
+            anyAllowed = false
+            for allowance in allowedSequences
+              anyAllowed = anyAllowed or allowance
+            if not anyAllowed
+              allowedSequences.fill(true)
+            else
+              for i in [0..curSequenceIndex - 1]
+                moveFromBoundary()
+
+            curSequence = null
+            curSequenceId = null
+            curSequenceIndex = null
+            curEquationOuterBoundary = null
+            curEquationInnerBoundary = null
+
+        if curSequence == null
```

Comment:
Pls add a comment to this if. What does it mean that `curSequence` is null and why do we do what is written below (scan all sequences again)